src/GenSequence.ts

```diff
@@ -282,6 +286,33 @@ export function first<T>(fn: (t: T) => boolean, defaultValue: T, i: Iterable<T>)
     return defaultValue;
 }
 
+export function max<T>(i: Iterable<T>): Maybe<T> {
+    let sawValue: boolean = false;
+    const value: Maybe<T> = reduce((p: T, c: T) => {
+        if (!sawValue) {
+            sawValue = true;
+        }
+        
+        if (isNonValue(p) || c > p) {
+            return c;
+        }
+
+        return p;
+    },
+    undefined, 
+    i);
+
+    if (!sawValue) {
+        throw new Error("the sequence is empty.");
+    }
+    
+    if (isNonValue(value)) {
+        return undefined;
+    }
+
+    return value;
+}
+
```

Comment:
Hmm, ok I implemented it the very clean way you mentioned, and it looks great, however the tests expose that it works normally if you start with a real value, but if you start with a non value (NaN or undefined) first, the non value gets "stuck" and that is what you end up getting.  I fear that folks will try it with non values in the sequence and get the impression that they are ignored, but then find themselves with a non value in the starting position in production and have it fail them.
```javascript
    it('test max starts with NaN always NaN', () => {
        const values = [NaN, 1, NaN, 2];
        expect(genSequence(values).max()).to.be.NaN;
    });

    it('test max NaN value', () => {
        const values = [1, NaN, 2];
        expect(genSequence(values).max()).to.equal(2);
    });
```
Are you ok leaving this slightly misleading behavior?  I can be ok with it.